src/object.rs

```diff
@@ -0,0 +1,321 @@
+                }
+
+                let arg = &*Rc::clone(args.first().unwrap());
+                match arg {
+                    Object::Array(a) => {
+                        match a.elements.last() {
+                            Some(el) => Ok(Rc::clone(el)),
+                            None => Ok(Rc::new(Object::Null)),
+                        }
+                    },
+                    obj => Err(format!("object {:?} not supported as an argument for last", obj))
+                }
+            },
+            Builtin::Rest => {
+                if args.len() != 1 {
+                    return Err("rest takes only 1 array argument".to_string())
+                }
+
+                let arg = &*Rc::clone(args.first().unwrap());
+                match arg {
+                    Object::Array(a) => {
+                        if a.elements.len() <= 1 {
+                            Ok(Rc::new(Object::Array(Rc::new(Array{elements: vec![]}))))
+                        } else {
+                            let mut elements = a.elements.clone();
+                            elements.remove(0);
+                            Ok(Rc::new(Object::Array(Rc::new(Array{elements}))))
+                        }
+                    },
+                    obj => Err(format!("object {:?} is not supported as an argument for rest", obj))
+                }
+            },
+            Builtin::Push => {
+                if args.len() != 2 {
+                    return Err("push takes an array and an object".to_string())
+                }
+
+                let array = &*Rc::clone(args.first().unwrap());
+                let obj = Rc::clone(args.last().unwrap());
+
+                // TODO: handle pushing objects like an array onto an array
+                match array {
+                    Object::Array(a) => {
+                        let mut elements = a.elements.clone();
+                        elements.push(obj);
+                        Ok(Rc::new(Object::Array(Rc::new(Array{elements}))))
+                    },
+                    _ => Err("first argument to push must be an array".to_string())
+                }
+            },
+            Builtin::Puts => {
+                for arg in args {
+                    println!("{}", arg.inspect())
+                }
+                Ok(Rc::new(Object::Null))
+            }
+        }
+    }
+
+    fn inspect(&self) -> String {
+        match self {
+            Builtin::Len => "len".to_string(),
+            Builtin::First => "first".to_string(),
+            Builtin::Last => "last".to_string(),
+            Builtin::Rest => "rest".to_string(),
+            Builtin::Push => "push".to_string(),
+            Builtin::Puts => "puts".to_string(),
+        }
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct Function {
+    pub parameters: Vec<ast::IdentifierExpression>,
+    pub body: ast::BlockStatement,
+    pub env: Rc<RefCell<Environment>>,
```

Comment:
This is the spot where I end up creating a memory leak since there are circular references. I'll log an issue to talk about what to do about this since it's a longer thing. One simple question though: I've seen `Rc<RefCell<...` and I've seen `RefCell<Rc<...`. Is there a difference between the two? Which is "more correct"?

Reply:
For avoiding memory leaks caused by cycles in the interpreted code, this gc library seems like a good fit (but it's in an early stage):
https://www.reddit.com/r/rust/comments/9ozaut/shifgrethor_i_garbage_collection_as_a_rust_library/

RefCell is like Cell (allows mutation) except it works for types that aren't `Copy`. Rc wraps it in a reference counted "box", so if you want to be able to mutate the same object from different (non-concurrent) code paths (in the same thread), you'd usually use `Rc<RefCell<T>>`, e.g. with the job_scheduler crate if you want to be able to mutate an object from different jobs (that are not run in parallel).
It's used when the borrow checker won't let you do it without this, because it can't prove that the mutable borrows don't overlap in time, but you know it for sure (RefCell will panic at runtime if a second mutable borrow happens while the first one is still active). 
This pattern is basically runtime borrow checking. It should only be used when compile-time borrow checking doesn't suffice (because there is the (small) ref counting overhead and it's better to prove the absence of a bug at compile-time than to check for it at runtime).
This pattern is also used heavily in the nphysics engine (the last time I checked) because it has to store a lot of cross references that allow mutating the objects.

AFAIR, I've never used the second way (`RefCell<Rc<...`) (since I started using Rust in 2014), where did you see that?

Reply:
I thought I had seen it the second way somewhere, but now I can't seem to dig up where I saw this. So maybe just misremembering all the different things I read on these.